asynclocalstorage.md

```diff
@@ -0,0 +1,371 @@
+
+The `unhandledrejection` and `rejectionhandled` events should, by default,
+propagate the context that was current when the promise was rejected and *not*
+when the promise was created.
+
+(Note that these are the only standard `EventTarget` events that are expected
+to propagate async context currently. All other events should be invoked in
+the same frame that the `dispatchEvent()` method was called in.)
+
+The deferred promise pattern is an approach to using promises that separates
+the `resolve()` and `reject()` functions used to settle the promise from the
+actual promise itself.
+
+```js
+function deferredPromise() {
+  let resolve, reject;
+  const promise = new Promise((a,b) => {
+    resolve = a;
+    reject = b;
+  });
+  return { promise, resolve, reject }
+}
+```
+
+Care must be taken when using this pattern to properly propagate the async
+context to the Web API `unhandledrejection` and `rejectionhandled` events.
+
+Specifically, given the following example, the `unhandledrejection` event
+should, by default, get the value `321` from the `AsyncLocalStorage`
+instance, rather than `123`:
+
+```js
+const als = new AsyncLocalStorage();
+
+addEventListener('unhandledrejection', (event) => {
+  console.log(als.getStore()); //321
+});
+
+const { promise, reject } = als.run(123, () => deferredPromise());
+als.run(321, () => reject('boom'));
+```
+
+If someone really wants the `unhandledrejection` event to receive the
+context at the time the promise was created, then the `reject()` should be
+wrapped using `AsyncResource.bind()`, for instance:
+
+```js
+function deferredPromiseCurrentContext() {
+  let resolve, reject;
+  const promise = new Promise((a,b) => {
+    resolve = AsyncResource.bind(a);
+    reject = AsyncResource.bind(b);
+  });
+  return { promise, resolve, reject }
+}
+```
+
+This would ensure that whenever `reject()` is called, it enters the
+context that was current when the promise was created.
+
+#### Thenables
+
+Thenables (promise-like objects that expose a `then()` method, should be handled
+in exactly the same way as a promise, by default.
+
+Thenables do, however, have the option of being defined as extending
+`AsyncResource`, overriding the default behavior.
+
+### queueMicrotask
+
+`queueMicrotask()` must capture the async context frame that is current when
+`queueMicrotask()` is called and run the callback within the context of that
+frame.
+
+### Timers
+
+Timers (using `setTimeout()`, `setInterval()`) must capture the async context
+frame that is current when the timer is created and must run the callback
+function within the context of that frame.
+
+### General guidelines for runtimes and applications
+
+The common theme is that the async context should be captured at the moment
+an asynchronous task is *scheduled*.
+
+  * `promise.then(fn)` *schedules* a continuation task
+  * `setTimeout(fn, n)` *schedules* a timer task
+  * `queueMicrotask(fn)` *schedules* a microtask
+  * and so on
```

Comment:
This paragraph and bullets explain the intuition really well, and could be shifted up a bit.